src/bin/ipa_bench/gen_events.rs

```diff
@@ -203,109 +80,104 @@ pub fn generate_events<R: RngCore + CryptoRng, W: io::Write>(
-                params.epoch,
-                u32::try_from(t.as_secs()).unwrap(),
-                params.breakdown_key,
-            )));
-        }
+    let matchkey = rng.gen::<MatchKey>();
+
+    // Randomly choose a datetime (plus the given base timestamp) as the first impression
+    let mut last_impression =
+        EventTimestamp::new(epoch, rng.gen_range(0..EventTimestamp::SECONDS_IN_EPOCH));
 
-        last_impression = t;
+    for _ in 0..impressions {
+        let ts = EventTimestamp::from(sample.impressions_time_diff(rng).as_secs());
+        last_impression = add_event_timestamps(last_impression, ts);
+
+        reports.push(GenericReport::Source {
+            event: Event {
+                matchkey,
+                attribution_constraint_id: None,
+                timestamp: last_impression,
+            },
+            breakdown_key,
+        });
     }
 
     // TODO: How should we simulate a case where there are multiple conversions and impressions in between? e.g. i -> i -> c -> i -> c
 
     let mut last_conversion = last_impression;
 
-    for _ in 0..params.conversions {
+    for _ in 0..conversions {
         let conversion_value = sample.conversion_value_per_ad(rng);
-        let t = last_conversion + sample.conversions_time_diff(rng);
-
-        if secret_share {
-            events.push(Event::EncryptedTrigger(ETriggerEvent {
-                event: EEvent {
-                    matchkeys: ss_mks.clone(),
-                    //TODO: Carry to next epoch if timestamp > DAYS_IN_EPOCH
-                    epoch: params.epoch,
-                    timestamp: u32::try_from(t.as_secs()).unwrap().xor_split(ss_rng),
-                },
-                value: conversion_value.xor_split(ss_rng),
-            }));
-        } else {
-            events.push(Event::Trigger(TriggerEvent::new(
-                matchkeys.clone(),
-                //TODO: Carry to next epoch if timestamp > DAYS_IN_EPOCH
-                params.epoch,
-                u32::try_from(t.as_secs()).unwrap(),
-                conversion_value,
-            )));
-        }
-
-        last_conversion = t;
+        let ts = EventTimestamp::from(sample.conversions_time_diff(rng).as_secs());
+        last_conversion = add_event_timestamps(last_conversion, ts);
+
+        reports.push(GenericReport::Trigger {
+            event: Event {
+                matchkey,
+                attribution_constraint_id: None,
+                timestamp: last_conversion,
+            },
+            value: conversion_value,
+        });
     }
 
-    events
+    reports
 }
 
-fn gen_matchkeys<R: RngCore + CryptoRng>(count: u8, rng: &mut R) -> Vec<MatchKey> {
-    let mut mks = Vec::new();
-
-    for _ in 0..count {
-        mks.push(rng.gen::<u64>());
+/// Adds two `EventTimestamp` instances.
+/// `offset` overflow will carry to `epoch`.
+/// `epoch` overflow will wrap and has no effect on `offset`.
+///
+/// # Exmaples
+///
+/// ```no_run
```

Comment:
I don't think it is discoverable by rust engine anyway :( I wasn't able to run it even with `no_run` removed